upgradeability_with_vtable/contracts/UpgradeabilityStorage.sol

```diff
@@ -0,0 +1,24 @@
+pragma solidity ^0.4.18;
+
+import './IRegistry.sol';
+
+/**
+ * @title UpgradeabilityStorage
+ * @dev This contract holds all the necessary state variables to support the upgrade functionality
+ */
+contract UpgradeabilityStorage {
+  // Registry of versions
+  IRegistry internal registry;
+
+  // Mapping of functions to current implementations
+  mapping (bytes4 => address) internal _implementations;
```

Comment:
+1 to caching the implementations hash locally in the proxy, to remove a call to the registry for each call to the proxy